### ISSUE TITLE
Add a migration page for "Migrating from StateNotifier to Async/Notifier"

### ISSUE THREAD

**Issue by rrousselGit** [rrousselGit/riverpod#2630]
This is a meta issue for tracking documentation work related to migrating from StateNotifier to Async/Notifier

Cf https://github.com/rrousselGit/riverpod/issues/2504#issuecomment-1584674828
Cf https://github.com/rrousselGit/riverpod/issues/1767

----

- [ ] Cross-link this document in the Provider migration guide, in the section talking about "Use ChangeNotifierProvider at first"
  cf https://github.com/rrousselGit/riverpod/pull/2676/files#r1260743192

- [ ] `StateNotifier.addListener` and `.stream` aren't present inside `Notifier`.
   If you want to listen to a `Notifier`, you can use `providerContainer.listen(provider)` or `ref.listen(provider)`.

- [ ] `StateNotifier.debugState` no longer exists.
  Tests which want to interact with Notifiers should always do so by reading the provider instead of directly instantiating the Notifier by hand.
  You can do `providerContainer.read(provider)` to get the current state of a provider inside tests.

- [ ] Document that Async/NotifierProviders do not receive a "ref" parameter.
       Logic that you would typically place inside the provider should be moved to the Notifier.build method, which has access to both `ref` and potential `family` parameters.
       Cf https://github.com/rrousselGit/riverpod/issues/2460

- [ ] Document that the Async/Notifier type is bound to autoDispose/family features. Enabling/disabling those notifiers means you should change the subclassed notifier type:
         | Provider                                                                          | Notifier                                                           |
    |-----------------------------------------------------------------------------------|--------------------------------------------------------------------|
    | NotifierProvider<MyNotifier, MyState>(MyNotifier.new)                             | class MyNotifier extends Notifier<MyState> {                       |
    | NotifierProvider.autoDispose<MyNotifier, MyState>(   MyNotifier.new)              | class MyNotifier extends Notifier<MyState> {                       |
    | NotifierProvider.autoDispose.family<MyNotifier, MyState, Arg>(   MyNotifier.new ) | class MyNotifier extends AutoDisposeFamilyNotifier<MyState, Arg> { |
    | NotifierProvider.family<MyNotifier, MyState, Arg>(   MyNotifier.new )             | class MyNotifier extends FamilyNotifier<MyState, Arg> {            |
    
    Note: AutoDispose variants will likely disappear once https://github.com/rrousselGit/riverpod/issues/2576 lands

- [ ] Document how Notifiers are not disposed on ref.watch/invalidate/...
  As such, ref.watch(provider.notifier) never emits a new value.
  This enables keeping state across provider rebuilds. For example, we can do:
  ```dart
  class MyNotifier extends Notifier<int> {
     bool _isFirstBuild = true;
    @override
    int build() {
      if (_isFirstBuild) {
        // TODO maybe read state from the local database
        _isFirstBuild = false;
      }
    }
  }
  ```
  This also means that instead of overriding `dispose`, notifiers should instead rely on `ref.onDispose`.
        

- [ ] Due to the difference in dispose life-cycle, there is no-longer a "mounted" property as a boolean wouldn't quite work.
        Related features: https://github.com/rrousselGit/riverpod/issues/2668 & https://github.com/rrousselGit/riverpod/issues/1660

- [ ] How to override "dispose"? Instead use `ref.onDispose`

- [ ] How to run logic after "build"
  ```dart
  class QrScanner extends StateNotifier<QrScannerState> {
    QrScanner() : super(const QrScannerState()) {
      _startScanner();
    }
  }
  ```
  vs:
  ```dart
  @riverpod
  class QrScanner extends _$QrScanner {
    @override
    QrScannerState build() {
     state = const QrScannerState();
     _startScanner();
     return state;
    }
  }
  ```

**Comment by nilsreichardt** [comment]
It would be great to mention in this page (or a different), that you can pass parameters to the build method (see https://github.com/rrousselGit/riverpod/discussions/2628#discussioncomment-6133853)

**Comment by lucavenir** [comment]
I think it's best to write a migration page "From Riverpod 1.0 to Riverpod 2.0".

That's when `Notifier` was introduced. And also other stuff happened as well. Is this being tracked? Or can we change this issue to be wider and track down this whole topic, instead?

My intention is to fill in another PR about this topic, so that I can cross reference it into #2676 and submit it.

**Comment by rrousselGit** [comment]
I'd be skeptical about the SEO of a 1=>2 page

What else would you include in it?

**Comment by lucavenir** [comment]
If I'm not mistaken, with v2 we obtained:

1. `Notifier` and `AsyncNotifier`
2. A different behavior when using `AsyncValue.when`: you still have the previous value or error when `AsyncValue` is loading or refreshing
3. The whole set of utilities connected to the above, such `valueOrNull`, `hasError`, `hasData`, `asError`, `isLoading`, etc.
4. Any other breaking change I might miss in here (I recall that `ProviderListener` was removed, for instance)

Overall it was a big change that justifies the deprecation of `StateNotifierProvider` and then `StateProvider` also.
